File: backend/routes/customers.py
# Customer Management Routes
from flask import Blueprint, request, jsonify
from services.supabase_client import get_supabase
from datetime import datetime

bp = Blueprint('customers', __name__)
supabase = get_supabase()

@bp.route('', methods=['GET'])
def get_customers():
    """Get all customers"""
    try:
        # Query parameters for search/filter
        search = request.args.get('search', '')
        business_type = request.args.get('business_type', '')
        
        query = supabase.table('customers').select('*')
        
        if search:
            query = query.or_(f'name.ilike.%{search}%,phone.ilike.%{search}%')
        
        if business_type:
            query = query.eq('business_type', business_type)
        
        response = query.order('created_at', desc=True).execute()
        
        return jsonify({
            'success': True,
            'data': response.data,
            'count': len(response.data)
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<customer_id>', methods=['GET'])
def get_customer(customer_id):
    """Get single customer by ID"""
    try:
        response = supabase.table('customers').select('*').eq('id', customer_id).execute()
        
        if not response.data:
            return jsonify({'success': False, 'error': 'Customer not found'}), 404
        
        return jsonify({
            'success': True,
            'data': response.data[0]
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('', methods=['POST'])
def create_customer():
    """Create new customer"""
    try:
        data = request.get_json()
        
        # Validate required fields
        required_fields = ['name', 'phone']
        for field in required_fields:
            if field not in data:
                return jsonify({'success': False, 'error': f'{field} is required'}), 400
        
        # Check if phone already exists
        existing = supabase.table('customers').select('id').eq('phone', data['phone']).execute()
        if existing.data:
            return jsonify({'success': False, 'error': 'Phone number already exists'}), 400
        
        # Insert customer
        customer_data = {
            'name': data['name'],
            'phone': data['phone'],
            'business_type': data.get('business_type', ''),
            'address': data.get('address', '')
        }
        
        response = supabase.table('customers').insert(customer_data).execute()
        
        return jsonify({
            'success': True,
            'message': 'Customer created successfully',
            'data': response.data[0]
        }), 201
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<customer_id>', methods=['PUT'])
def update_customer(customer_id):
    """Update customer"""
    try:
        data = request.get_json()
        
        # Check if customer exists
        existing = supabase.table('customers').select('id').eq('id', customer_id).execute()
        if not existing.data:
            return jsonify({'success': False, 'error': 'Customer not found'}), 404
        
        # Update customer
        update_data = {}
        if 'name' in data:
            update_data['name'] = data['name']
        if 'phone' in data:
            update_data['phone'] = data['phone']
        if 'business_type' in data:
            update_data['business_type'] = data['business_type']
        if 'address' in data:
            update_data['address'] = data['address']
        
        response = supabase.table('customers').update(update_data).eq('id', customer_id).execute()
        
        return jsonify({
            'success': True,
            'message': 'Customer updated successfully',
            'data': response.data[0]
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<customer_id>', methods=['DELETE'])
def delete_customer(customer_id):
    """Delete customer"""
    try:
        # Check if customer exists
        existing = supabase.table('customers').select('id').eq('id', customer_id).execute()
        if not existing.data:
            return jsonify({'success': False, 'error': 'Customer not found'}), 404
        
        # Delete customer
        supabase.table('customers').delete().eq('id', customer_id).execute()
        
        return jsonify({
            'success': True,
            'message': 'Customer deleted successfully'
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/search', methods=['GET'])
def search_by_phone():
    """Search customer by phone number"""
    try:
        phone = request.args.get('phone', '')
        
        if not phone:
            return jsonify({'success': False, 'error': 'Phone number is required'}), 400
        
        response = supabase.table('customers').select('*').eq('phone', phone).execute()
        
        if not response.data:
            return jsonify({
                'success': True,
                'found': False,
                'data': None
            }), 200
        
        # Get customer order history
        customer_id = response.data[0]['id']
        orders = supabase.table('orders').select('*').eq('customer_id', customer_id).order('order_date', desc=True).limit(5).execute()
        
        return jsonify({
            'success': True,
            'found': True,
            'data': {
                'customer': response.data[0],
                'recent_orders': orders.data
            }
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

File: backend-node/routes/customers.js
// Customers Routes
const express = require('express');
const router = express.Router();
const store = require('../data/store');

// GET /api/customers - List all customers
router.get('/', (req, res) => {
    try {
        let customers = store.findAll('customers');

        // Filter by search (name or phone)
        if (req.query.search) {
            const search = req.query.search.toLowerCase();
            customers = customers.filter(
                c => c.name.toLowerCase().includes(search) || c.phone.includes(search)
            );
        }

        // Filter by business type
        if (req.query.business_type) {
            customers = customers.filter(c => c.business_type === req.query.business_type);
        }

        res.json({ success: true, data: customers });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// GET /api/customers/search?phone= - Search by phone
router.get('/search', (req, res) => {
    try {
        const { phone } = req.query;
        if (!phone) {
            return res.status(400).json({ success: false, error: 'Phone number is required' });
        }

        const customer = store.findAll('customers').find(c => c.phone === phone);

        if (customer) {
            res.json({ success: true, found: true, data: { customer } });
        } else {
            res.json({ success: true, found: false, data: null });
        }
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// GET /api/customers/:id - Get customer by ID
router.get('/:id', (req, res) => {
    try {
        const customer = store.findById('customers', req.params.id);
        if (!customer) {
            return res.status(404).json({ success: false, error: 'Customer not found' });
        }
        res.json({ success: true, data: customer });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// POST /api/customers - Create customer
router.post('/', (req, res) => {
    try {
        const { name, phone, business_type, address } = req.body;

        if (!name || !phone) {
            return res.status(400).json({ success: false, error: 'Name and phone are required' });
        }

        // Check duplicate phone
        const existing = store.findAll('customers').find(c => c.phone === phone);
        if (existing) {
            return res.status(400).json({ success: false, error: 'Customer with this phone already exists' });
        }

        const customer = store.create('customers', { name, phone, business_type: business_type || '', address: address || '' });
        res.status(201).json({ success: true, data: customer });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// PUT /api/customers/:id - Update customer
router.put('/:id', (req, res) => {
    try {
        const updated = store.update('customers', req.params.id, req.body);
        if (!updated) {
            return res.status(404).json({ success: false, error: 'Customer not found' });
        }
        res.json({ success: true, data: updated });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// DELETE /api/customers/:id - Delete customer
router.delete('/:id', (req, res) => {
    try {
        const deleted = store.remove('customers', req.params.id);
        if (!deleted) {
            return res.status(404).json({ success: false, error: 'Customer not found' });
        }
        res.json({ success: true, message: 'Customer deleted' });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

module.exports = router;
